### ISSUE TITLE
Kubernetes CronJob pods is not getting clean-up when Job is completed

### ISSUE THREAD

**Issue by dcvtruong** [kubernetes/kubernetes#74741]
sig/apps

**What happened**:
I've created a CronJob with setting failedJobsHistoryLimit at 1 and successfulJobsHistoryLimit at 3. The pod spawned from the jobs is finished with status 'Completed' but not getting deleted from each scheduling so the number of Pods in the k8s cluster keeps increasing.
**What you expected to happen**:
The pods spawned from the cronjob should get cleanup after the job has completed
**How to reproduce it (as minimally and precisely as possible)**:
Deploy elasticsearch curator 5.5.4 (stable/elasticsearch-curator)
**Anything else we need to know?**:

       # kubectl get cronjob -o yaml

			apiVersion: batch/v1beta1
			kind: CronJob
			metadata:
			  labels:
			    app: elasticsearch-curator
			    chart: elasticsearch-curator-1.0.1
			    heritage: Tiller
			    release: curator
			  name: curator-elasticsearch-curator
			  namespace: elk
			spec:
			  concurrencyPolicy: Allow
			  failedJobsHistoryLimit: 1
			  jobTemplate:
			    metadata:
			      creationTimestamp: null
			      labels:
			        app: elasticsearch-curator
			        release: curator
			    spec:
			      template:
			        metadata:
			          creationTimestamp: null
			          labels:
			            app: elasticsearch-curator
			            release: curator
			        spec:
			          containers:
			          - args:
			            - --config
			            - /etc/es-curator/config.yml
			            - /etc/es-curator/action_file.yml
			            command:
			            - curator
			            image: quay.io/pires/docker-elasticsearch-curator:5.5.4
			            imagePullPolicy: IfNotPresent
			            name: elasticsearch-curator
			            resources:
			              limits:
			                cpu: 300m
			                memory: 256Mi
			              requests:
			                cpu: 100m
			                memory: 128Mi
			            terminationMessagePath: /dev/termination-log
			            terminationMessagePolicy: File
			            volumeMounts:
			            - mountPath: /etc/es-curator
			              name: config-volume
			          dnsPolicy: ClusterFirst
			          restartPolicy: Never
			          schedulerName: default-scheduler
			          securityContext: {}
			          terminationGracePeriodSeconds: 30
			          volumes:
			          - configMap:
			              defaultMode: 420
			              name: curator-elasticsearch-curator-config
			            name: config-volume
			  schedule: */59 * * * *
			  successfulJobsHistoryLimit: 3
			  suspend: false

       # kubectl get pod -n elk

                 NAMESPACE        NAME                                                        READY   STATUS    RESTARTS     AGE
                    elk           curator-elasticsearch-curator-1550807940-94vcr              0/1     Completed   0          12h
			elk           curator-elasticsearch-curator-1550808000-7dj8l              0/1     Completed   0          12h
			elk           curator-elasticsearch-curator-1550811540-7kpnt              0/1     Completed   0          11h
			elk           curator-elasticsearch-curator-1550811600-r4sql              0/1     Completed   0          11h
			elk           curator-elasticsearch-curator-1550815140-rrmdv              0/1     Completed   0          10h
			elk           curator-elasticsearch-curator-1550815200-h6v47              0/1     Completed   0          10h
			elk           curator-elasticsearch-curator-1550818740-wgm8n              0/1     Completed   0          9h
			elk           curator-elasticsearch-curator-1550818800-6pr9f              0/1     Completed   0          9h
			elk           curator-elasticsearch-curator-1550822340-qg7qg              0/1     Completed   0          8h
			elk           curator-elasticsearch-curator-1550822400-dgh5v              0/1     Completed   0          8h
			elk           curator-elasticsearch-curator-1550825940-mxf4p              0/1     Completed   0          7h47m
			elk           curator-elasticsearch-curator-1550826000-8wxt9              0/1     Completed   0          7h46m
			elk           curator-elasticsearch-curator-1550829540-2bjfq              0/1     Completed   0          6h47m
			elk           curator-elasticsearch-curator-1550829600-tg8qj              0/1     Completed   0          6h46m
			elk           curator-elasticsearch-curator-1550833140-xt5vp              0/1     Completed   0          5h47m
			elk           curator-elasticsearch-curator-1550833200-wz996              0/1     Completed   0          5h46m
			elk           curator-elasticsearch-curator-1550836740-tdhg2              0/1     Completed   0          4h47m
			elk           curator-elasticsearch-curator-1550836800-96mz6              0/1     Completed   0          4h46m
			elk           curator-elasticsearch-curator-1550840340-ljz2c              0/1     Completed   0          3h47m
			elk           curator-elasticsearch-curator-1550840400-pftl4              0/1     Completed   0          3h46m
			elk           curator-elasticsearch-curator-1550843940-27pdn              0/1     Completed   0          167m
			elk           curator-elasticsearch-curator-1550844000-vl452              0/1     Completed   0          166m
			elk           curator-elasticsearch-curator-1550847540-mrfbp              0/1     Completed   0          107m
			elk           curator-elasticsearch-curator-1550847600-tp2fv              0/1     Completed   0          106m
			elk           curator-elasticsearch-curator-1550851140-d4ggz              0/1     Completed   0          47m
			elk           curator-elasticsearch-curator-1550851200-cgr2m              0/1     Completed   0          46m

**Environment**:
- Kubernetes version (use `kubectl version`): v1.13
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): RHEL 7.6
- Kernel (e.g. `uname -a`): 3.10.0-957.1.3.el7.x86_64
- Install tools: kubespray
- Others: curator 5.5.4

**Comment by Joseph-Irving** [comment]
Can you paste the output of `kubectl get jobs` and `kubectl desribe cronjob 
curator-elasticsearch-curator`?

**Comment by dcvtruong** [comment]
Logs:

      #kubectl describe cronjob curator-elasticsearch-curator -n elk

      Name:                       curator-elasticsearch-curator
      Namespace:                  elk
      Labels:                     app=elasticsearch-curator
                                  chart=elasticsearch-curator-1.0.1
                                  heritage=Tiller
                                  release=curator
      Annotations:                <none>
      Schedule:                   0 */4 * * *
      Concurrency Policy:         Allow
      Suspend:                    False
      Starting Deadline Seconds:  <unset>
      Selector:                   <unset>
      Parallelism:                <unset>
      Completions:                <unset>
      Pod Template:
        Labels:  app=elasticsearch-curator
                 release=curator
        Containers:
         elasticsearch-curator:
          Image:      quay.io/pires/docker-elasticsearch-curator:5.5.4
          Port:       <none>
          Host Port:  <none>
          Command:
            curator
          Args:
            --config
            /etc/es-curator/config.yml
            /etc/es-curator/action_file.yml
          Environment:  <none>
          Mounts:
            /etc/es-curator from config-volume (rw)
        Volumes:
         config-volume:
          Type:            ConfigMap (a volume populated by a ConfigMap)
          Name:            curator-elasticsearch-curator-config
          Optional:        false
      Last Schedule Time:  Fri, 01 Mar 2019 12:00:00 +0000
      Active Jobs:         <none>
      Events:              <none>

**Comment by goodluckbot** [comment]
@dcvtruong 
seems your job runs twice every hour, do you have any error log with it
```
		elk           curator-elasticsearch-curator-1550851140-d4ggz              0/1     Completed   0          47m
		elk           curator-elasticsearch-curator-1550851200-cgr2m              0/1     Completed   0          46m
```
And also, your cronjob has 3 jobs, which means the other finished jobs are deleted. it is just the jobs are not deleted by time order. can you please also check your /var/log/message to see if there's any error when removing docker?

**Comment by dcvtruong** [comment]
I will get the log when the next curator job cycle.  Thanks.

**Comment by janetkuo** [comment]
It looks like the Jobs are cleaned up, but their Pods aren't garbage collected with them. I wonder if those Pods have finalizers set on them. 

@dcvtruong did you see finalizers being set in those Pods' `metadata`? Do they have `metadata.deletionTimestamp` set?

**Comment by dcvtruong** [comment]
Hi @janetkuo, There are no metadata.deletionTimestamp set.

      metadata:
        annotations:
          kubernetes.io/limit-ranger: 'LimitRanger plugin set: cpu, memory request for container
            elasticsearch-curator; cpu, memory limit for container elasticsearch-curator'
        creationTimestamp: "2019-04-08T20:00:09Z"
        generateName: curator-elasticsearch-curator-1554753600-
        labels:
          app: elasticsearch-curator
          controller-uid: e9897845-5a38-11e9-9f68-000d3a2fd48b
          job-name: curator-elasticsearch-curator-1554753600
          release: curator
        name: curator-elasticsearch-curator-1554753600-vtwfj
        namespace: elk
        ownerReferences:
        - apiVersion: batch/v1
          blockOwnerDeletion: true
          controller: true
          kind: Job

**Comment by dcvtruong** [comment]
Edited the Cronjob with the 'finalizers' did not seems to apply to the Job or Pod.  Any suggestion?

      apiVersion: batch/v1beta1
      kind: CronJob
      metadata:
        creationTimestamp: "2019-04-30T19:48:23Z"
        finalizers:
        - foregroundDeletion
        labels:
          app: elasticsearch-curator
          chart: elasticsearch-curator-1.0.1
          heritage: Tiller
          release: curator
        name: curator-elasticsearch-curator
        namespace: elk
        resourceVersion: "19178606"
        selfLink: /apis/batch/v1beta1/namespaces/elk/cronjobs/curator-elasticsearch-curator
        uid: ea1d3964-6b80-11e9-822b-000d3a2fdfb0
      spec:
        concurrencyPolicy: Allow
        failedJobsHistoryLimit: 2
        jobTemplate:
          metadata:
            creationTimestamp: null
            labels:
              app: elasticsearch-curator
              release: curator
          spec:
            template:
              metadata:
                creationTimestamp: null
                labels:
                  app: elasticsearch-curator
                  release: curator
              spec:
                containers:
                - args:
                  - --config
                  - /etc/es-curator/config.yml
                  - /etc/es-curator/action_file.yml
                  command:
                  - curator
                  image: quay.io/pires/docker-elasticsearch-curator:5.5.4
                  imagePullPolicy: IfNotPresent
                  name: elasticsearch-curator
                  resources: {}
                  terminationMessagePath: /dev/termination-log
                  terminationMessagePolicy: File
                  volumeMounts:
                  - mountPath: /etc/es-curator
                    name: config-volume
                dnsPolicy: ClusterFirst
                restartPolicy: Never
                schedulerName: default-scheduler
                securityContext: {}
                terminationGracePeriodSeconds: 30
                volumes:
                - configMap:
                    defaultMode: 420
                    name: curator-elasticsearch-curator-config
                  name: config-volume
        schedule: 0 */1 * * *
        successfulJobsHistoryLimit: 3
        suspend: false

**Comment by dcvtruong** [comment]
Looks like the finalizers is not supported in CronJob.  Also, it was set in the job template of the CronJob and did not work.

**Comment by Arziel** [comment]
Still nothing? There are lots of solved issues, but doesn't seems fixed at all.

#71772 #71796 #71975

**Comment by dcvtruong** [comment]
Now running K8s 1.13.10 and it seems the completed (cronjob) pods are not automatically remove.  I thought this was resolved in 1.13.3

**Comment by IdanAdar** [comment]
Pretty sure this is also the case in 1.14.x.

**Comment by tedyu** [comment]
```
	if err := jc.DeleteJob(job.Namespace, job.Name); err != nil {
		recorder.Eventf(sj, v1.EventTypeWarning, "FailedDelete", "Deleted job: %v", err)
		klog.Errorf("Error deleting job %s from %s: %v", job.Name, nameForLog, err)
```
Was there log in the above form ?

**Comment by antmonnier** [comment]
I'm using 1.16.0 and pods are still not removed after jobs are deleted.

**Comment by ryedin** [comment]
This appears to be true still (normal Jobs with `ttlSecondsAfterFinished: 0` don't get removed for me), so I have the job as well as the pod sitting there indefinitely

**Comment by tedyu** [comment]
@ryedin 
See if #85141 fixes the issue you described above.

**Comment by ryedin** [comment]
Sorry, not power user enough yet to test some arbitrary branch of the kube api. Looking at the code in the PR, I guess it seems good :) 

Looking at the diff led me to believe that I could maybe get around this issue by specifying a non-zero positive value, but that appears to also not work.

/shrug

fwiw, this is on `v1.14.7`

**Comment by ryedin** [comment]
I'm guessing my issue is more just how my current server (docker desktop with kubernetes enabled) has chosen to implement the garbage collection (or not).

we're in early phases of modernizing our app, so this is all the data I have right now (still in local testing, then will move to EKS/eksctl, etc).

Apologies if all I've done is add noise to this issue xD

**Comment by 2rs2ts** [comment]
I don't know if the shrug label deprioritizes this but if it does can we /unshrug cuz this issue is actually pretty nasty when it happens

**Comment by ryedin** [comment]
omfg I had no idea that would happen. TIL (+oops)

**Comment by dims** [comment]
/unshrug

:)

**Comment by soltysh** [comment]
This issue should've been fixed with https://github.com/kubernetes/kubernetes/pull/71801 (which is 1.14) can you check controller-manager logs (or events) for lines pointing that the job is being removed. Check https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L373 for what kind of logs/events you want to look for. Additionally I'd need to see if your garbage collector is processing those pods correctly. The logic from that PR enforces the pods to be removed by the garbage collector. Can you put yaml of one of the pods that was left behind like Janet asked earlier?

**Comment by ryedin** [comment]
@soltysh I'm unsure how to get the log/event data you are asking for, but here is the yaml from a pod that has been in the cluster for > 16 hours after the job completed. This is from installing the kubeapps helm chart yesterday while playing around.

The container terminates fine, of course, but the pod(s) and the job(s) just remain indefinitely.

```
metadata:
  name: apprepo-sync-bitnami-27pfx-6r2kr
  generateName: apprepo-sync-bitnami-27pfx-
  namespace: kubeapps
  selfLink: /api/v1/namespaces/kubeapps/pods/apprepo-sync-bitnami-27pfx-6r2kr
  uid: b215739a-065e-11ea-8c38-025000000001
  resourceVersion: '1072581'
  creationTimestamp: '2019-11-13T21:43:57Z'
  labels:
    apprepositories.kubeapps.com/repo-name: bitnami
    controller-uid: b213c9f7-065e-11ea-8c38-025000000001
    job-name: apprepo-sync-bitnami-27pfx
  ownerReferences:
    - apiVersion: batch/v1
      kind: Job
      name: apprepo-sync-bitnami-27pfx
      uid: b213c9f7-065e-11ea-8c38-025000000001
      controller: true
      blockOwnerDeletion: true
spec:
  volumes:
    - name: default-token-d6g6v
      secret:
        secretName: default-token-d6g6v
        defaultMode: 420
  containers:
    - name: sync
      image: 'docker.io/bitnami/kubeapps-chart-repo:1.9.0-r2'
      command:
        - /chart-repo
      args:
        - sync
        - '--mongo-url=kubeapps-mongodb'
        - '--mongo-user=root'
        - '--user-agent-comment=kubeapps/v1.6.1'
        - bitnami
        - 'https://charts.bitnami.com/bitnami'
      env:
        - name: MONGO_PASSWORD
          valueFrom:
            secretKeyRef:
              name: kubeapps-mongodb
              key: mongodb-root-password
      resources: {}
      volumeMounts:
        - name: default-token-d6g6v
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      imagePullPolicy: IfNotPresent
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 30
  dnsPolicy: ClusterFirst
  serviceAccountName: default
  serviceAccount: default
  nodeName: docker-desktop
  securityContext: {}
  schedulerName: default-scheduler
  tolerations:
    - key: node.kubernetes.io/not-ready
      operator: Exists
      effect: NoExecute
      tolerationSeconds: 300
    - key: node.kubernetes.io/unreachable
      operator: Exists
      effect: NoExecute
      tolerationSeconds: 300
  priority: 0
  enableServiceLinks: true
status:
  phase: Succeeded
  conditions:
    - type: Initialized
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2019-11-13T21:43:57Z'
      reason: PodCompleted
    - type: Ready
      status: 'False'
      lastProbeTime: null
      lastTransitionTime: '2019-11-13T21:44:55Z'
      reason: PodCompleted
    - type: ContainersReady
      status: 'False'
      lastProbeTime: null
      lastTransitionTime: '2019-11-13T21:44:55Z'
      reason: PodCompleted
    - type: PodScheduled
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2019-11-13T21:43:57Z'
  hostIP: 192.168.65.3
  podIP: 10.1.1.107
  startTime: '2019-11-13T21:43:57Z'
  containerStatuses:
    - name: sync
      state:
        terminated:
          exitCode: 0
          reason: Completed
          startedAt: '2019-11-13T21:43:58Z'
          finishedAt: '2019-11-13T21:44:55Z'
          containerID: >-
            docker://e87c5adcd1569b4c265748df9e6b4db80955d79b454ff2e432c50edefed68c47
      lastState: {}
      ready: false
      restartCount: 0
      image: 'bitnami/kubeapps-chart-repo:1.9.0-r2'
      imageID: >-
        docker-pullable://bitnami/kubeapps-chart-repo@sha256:c6ed3125daf7d16dd1a976bf155730903b1d8395807b6684acc235e89b1327a5
      containerID: >-
        docker://e87c5adcd1569b4c265748df9e6b4db80955d79b454ff2e432c50edefed68c47
  qosClass: BestEffort
kind: Pod

```

and again, this is on docker desktop, `kubectl version` output:

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.7", GitCommit:"8fca2ec50a6133511b771a11559e24191b1aa2b4", GitTreeState:"clean", BuildDate:"2019-09-18T14:47:22Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.7", GitCommit:"8fca2ec50a6133511b771a11559e24191b1aa2b4", GitTreeState:"clean", BuildDate:"2019-09-18T14:39:02Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by ryedin** [comment]
and the corresdponding job yaml:

```
metadata:
  name: apprepo-sync-stable-kfx5c
  generateName: apprepo-sync-stable-
  namespace: kubeapps
  selfLink: /apis/batch/v1/namespaces/kubeapps/jobs/apprepo-sync-stable-kfx5c
  uid: b2263ffe-065e-11ea-8c38-025000000001
  resourceVersion: '1072724'
  creationTimestamp: '2019-11-13T21:43:57Z'
  labels:
    apprepositories.kubeapps.com/repo-name: stable
    controller-uid: b2263ffe-065e-11ea-8c38-025000000001
    job-name: apprepo-sync-stable-kfx5c
  ownerReferences:
    - apiVersion: kubeapps.com/v1alpha1
      kind: AppRepository
      name: stable
      uid: b0a0b384-065e-11ea-8c38-025000000001
      controller: true
      blockOwnerDeletion: true
spec:
  parallelism: 1
  completions: 1
  backoffLimit: 6
  selector:
    matchLabels:
      controller-uid: b2263ffe-065e-11ea-8c38-025000000001
  template:
    metadata:
      creationTimestamp: null
      labels:
        apprepositories.kubeapps.com/repo-name: stable
        controller-uid: b2263ffe-065e-11ea-8c38-025000000001
        job-name: apprepo-sync-stable-kfx5c
    spec:
      containers:
        - name: sync
          image: 'docker.io/bitnami/kubeapps-chart-repo:1.9.0-r2'
          command:
            - /chart-repo
          args:
            - sync
            - '--mongo-url=kubeapps-mongodb'
            - '--mongo-user=root'
            - '--user-agent-comment=kubeapps/v1.6.1'
            - stable
            - 'https://kubernetes-charts.storage.googleapis.com'
          env:
            - name: MONGO_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: kubeapps-mongodb
                  key: mongodb-root-password
          resources: {}
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
      restartPolicy: OnFailure
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      securityContext: {}
      schedulerName: default-scheduler
status:
  conditions:
    - type: Complete
      status: 'True'
      lastProbeTime: '2019-11-13T21:46:51Z'
      lastTransitionTime: '2019-11-13T21:46:51Z'
  startTime: '2019-11-13T21:43:57Z'
  completionTime: '2019-11-13T21:46:51Z'
  succeeded: 1
kind: Job
```

and the results of `kubectl -n kubeapps describe job apprepo-sync-stable-kfx5c`:

```
Name:           apprepo-sync-stable-kfx5c
Namespace:      kubeapps
Selector:       controller-uid=b2263ffe-065e-11ea-8c38-025000000001
Labels:         apprepositories.kubeapps.com/repo-name=stable
                controller-uid=b2263ffe-065e-11ea-8c38-025000000001
                job-name=apprepo-sync-stable-kfx5c
Annotations:    <none>
Controlled By:  AppRepository/stable
Parallelism:    1
Completions:    1
Start Time:     Wed, 13 Nov 2019 15:43:57 -0600
Completed At:   Wed, 13 Nov 2019 15:46:51 -0600
Duration:       2m54s
Pods Statuses:  0 Running / 1 Succeeded / 0 Failed
Pod Template:
  Labels:  apprepositories.kubeapps.com/repo-name=stable
           controller-uid=b2263ffe-065e-11ea-8c38-025000000001
           job-name=apprepo-sync-stable-kfx5c
  Containers:
   sync:
    Image:      docker.io/bitnami/kubeapps-chart-repo:1.9.0-r2
    Port:       <none>
    Host Port:  <none>
    Command:
      /chart-repo
    Args:
      sync
      --mongo-url=kubeapps-mongodb
      --mongo-user=root
      --user-agent-comment=kubeapps/v1.6.1
      stable
      https://kubernetes-charts.storage.googleapis.com
    Environment:
      MONGO_PASSWORD:  <set to the key 'mongodb-root-password' in secret 'kubeapps-mongodb'>  Optional: false
    Mounts:            <none>
  Volumes:             <none>
Events:                <none>
```

not sure why I don't see any events here. 

I'll leave it to you experts to decide whether we chalk this up to "docker desktop kube sucks" for now.

**Comment by mtttcgcg** [comment]
We also see this behavior in 1.14.6.  Jobs are cleaned up, but the associated pods are not.

**Comment by JnMik** [comment]
Once the job is finished and the pod has been removed, is there a way to get the logs of the terminated pod ? Even if It's not the the list of _kubectl get pods_. anymore ?

**Comment by tedyu** [comment]
I couldn't find failed_jobs_history_limit in k8s codebase.
Once pod is removed, its logs are no longer kept.

**Comment by JnMik** [comment]
@tedyu I supposed it should if terraform offers it, na ? https://www.terraform.io/docs/providers/kubernetes/r/cron_job.html

**Comment by oldthreefeng** [comment]
We also see this behavior in 1.18.3.  Jobs are cleaned up, but the associated pods are not.

job

```
$  kubectl get jobs -n kube-system 
NAME                     COMPLETIONS   DURATION   AGE
backup-etcd-1600758000   1/1           10s        22h
```
job with pod.

```
$ kubectl get pod -n kube-system | grep bac
backup-etcd-1600758000-c6cb2               0/1     Completed   0          22h
```

cronjob

```
$ kubectl get cronjobs.batch -n kube-system 
NAME          SCHEDULE     SUSPEND   ACTIVE   LAST SCHEDULE   AGE
backup-etcd   0 15 * * *   False     0        22h             15d
```

yaml 

```
apiVersion: batch/v1beta1  
kind: CronJob 
metadata:
  name: backup-etcd 
  namespace: kube-system
spec: 
  successfulJobsHistoryLimit: 1
  failedJobsHistoryLimit: 1
  schedule: "0 15 * * *"  
  jobTemplate:
    spec:
      template:
        spec:
          restartPolicy: Never 
          nodeSelector:
            sealos: backup
          volumes:
          - hostPath:
              path: /root/.sealos
              type: DirectoryOrCreate
            name: sealos-home
          - configMap:
              items:
              - key: id_rsa 
                path: id_rsa
              name: sealos-pkflie
            name: sealos-pkflie
          containers:
          - name: sealos
            image: louisehong/sealos:3.3.10
            args:
            - etcd 
            - save
            - --docker
            volumeMounts:
            - mountPath: /root/.sealos 
              name: sealos-home
            - mountPath: /root/.ssh/id_rsa
              name: sealos-pkflie
              subPath: id_rsa
```

**Comment by elwood218** [comment]
Sorry but why is it so difficult to fix this??!??! This is a very annoying problem. If a job starts every minute then in few days the GKE quota is reached - Why this is not prioritised?? Why no one looks at this problem??

**Comment by dims** [comment]
There's work in progress, please follow along in https://github.com/kubernetes/kubernetes/issues/74741 and help test it to ensure the same problem does not occur there as well.

https://github.com/kubernetes/kubernetes/issues/74741

/close

-- Dims

**Comment by IdanAdar** [comment]
You’re linking to this very ticket @dims what is the correct ticket to follow?

**Comment by dims** [comment]
Fixed @IdanAdar

**Comment by throrin19** [comment]
Same problem Here. If it's fixed, what is the modification to make to our cronJobs ?

**Comment by themightychris** [comment]
It looks like the referenced fix was merged into 1.20 -- is there any known solution for 1.18 users?

**Comment by sabattle** [comment]
We're currently running k8s 1.19 and we're running into this issue. Is there a known solution for 1.19.x users?

**Comment by ravitiwari87** [comment]
Is the issue fixed ?

**Comment by bwinter** [comment]
On 1.22 and we are still seeing this as well.

**Comment by jurabek** [comment]
same problem, open the issue please

**Comment by b0b0haha** [comment]
It seems in cronjob controller v2, it didn't list the pods that related to the job that is going to be deleted, instead it set PropagationPolicy as metav1.DeletePropagationBackground and call the Delete client-go API directly.
![image](https://user-images.githubusercontent.com/28356082/190130249-2e7bd471-12f4-4dc5-acb7-92e1a4c01f41.png)
![image](https://user-images.githubusercontent.com/28356082/190130312-59bc69c6-0592-45e4-93a1-3509a933f44b.png)
![image](https://user-images.githubusercontent.com/28356082/190130336-86183a92-52cd-414e-9345-9de2cf079f00.png)

the delete Storage impl is as below:
![image](https://user-images.githubusercontent.com/28356082/190130478-d6a7ce39-7eda-4770-86c9-4b22d80e9bc9.png)

it calls deletionFinalizersForGarbageCollection, the question is whether the cronjob support finalizer?

**Comment by jibinbabu** [comment]
is this resolved? My cron job pods doesn't seem to garbage collect. This is one 1.24.x

**Comment by davidgiga1993** [comment]
> I try on 1.27.x and this issue still exist.

We're facing the same issues on 1.27 (AWS EKS). The jobs are correctly cleaned, the pods are kept forever. As a workaround I've built a cleanup utility to remove any pods having a parent job which doesn't exist anymore: https://github.com/davidgiga1993/cronjob-pod-cleaner

**Comment by abhihome-github** [comment]
I could solve this by setting 
successfulJobsHistoryLimit: 0
failedJobsHistoryLimit: 0

**Comment by Jinkxed** [comment]
Still a problem on 1.27 AWS EKS:
`failedJobsHistoryLimit` set to 1 and we have 7 in an error state.

**Comment by anuroop08** [comment]
> Still a problem on 1.27 AWS EKS: `failedJobsHistoryLimit` set to 1 and we have 7 in an error state.

Any luck ? facing same issue

**Comment by mmcossu** [comment]
same here for AKS `"kubernetesVersion": "1.28.14"`

**Comment by 2rs2ts** [comment]
May or may not be related to this, but we had a CronJob on some old 1.25 cluster that had a Job that didn't realize its Pod had finished running nearly 24 hours after the fact, so it basically had the same end result as this issue, just, it seemed like the confusion _might_ have been in the Job control logic, rather than the CronJob logic. We could not figure out what happened to make it act this way, based on server logs; all we knew was that the Pod was in a Completed (Successful) state, but the Job thought it was still running and the CronJob thus rightfully saw a running Job and didn't start a new Job (concurrent runs forbidden obv.)

Only commented this because maybe that's what people above are experiencing too, not sure though. Having an issue closed for years but still seemingly affecting people is quite frustrating, but if it's actually a separate-but-related bug, then maybe opening a new issue is feasible.

**Comment by mhemken-vts** [comment]
I'm on EKS `1.29`. Still an issue.

**Comment by srana6** [comment]
We are also seeing this issue in our cluster. I swear it was working fine and it broke when we upgraded to EKS v1.27

**Comment by dims** [comment]
For those of you reporting EKS issues here, it is not useful! please open a support ticket with AWS.
